File: app/src/main/java/com/algokelvin/tiktokcoinmanager/domain/model/Transaction.kt
package com.algokelvin.tiktokcoinmanager.domain.model

import java.time.Instant

data class Transaction(
    val id: String,
    val title: String,
    val transactionType: TransactionType,
    val coinAmount: Long,
    val topupExpense: Long?,
    val currency: String,
    val note: String,
    val transactionDate: Instant,
    val createdAt: Instant,
    val updatedAt: Instant,
)

enum class TransactionType {
    CREDIT,
    DEBIT,
}

File: backend/internal/domain/transaction.go
package domain

import "time"

type TransactionType string

const (
	TransactionTypeCredit TransactionType = "CREDIT"
	TransactionTypeDebit  TransactionType = "DEBIT"
	DefaultCurrency       string          = "IDR"
)

type Transaction struct {
	ID              string
	Title           string
	TransactionType TransactionType
	CoinAmount      int64
	TopupExpense    *int64
	Currency        string
	Note            string
	TransactionDate time.Time
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

func (t Transaction) Validate() error {
	if t.TransactionType != TransactionTypeCredit && t.TransactionType != TransactionTypeDebit {
		return ErrInvalidInput
	}
	if t.CoinAmount <= 0 {
		return ErrInvalidInput
	}
	if t.TopupExpense != nil && *t.TopupExpense < 0 {
		return ErrInvalidInput
	}
	if t.Currency == "" {
		t.Currency = DefaultCurrency
	}
	if t.TransactionDate.IsZero() {
		return ErrInvalidInput
	}
	return nil
}
